src/bindings/python/flux/job/output.py

```diff
@@ -0,0 +1,957 @@
+    """
+    jobid = JobID(jobid)
+    if nowait:
+        # Do not wait, fetch current output eventlog if available and
+        # return JobOutput object
+        eventlog = job_kvs_lookup(flux_handle, jobid, keys=["guest.output"])
+        if eventlog is None:
+            msg = f"job {jobid} does not exist or output not ready"
+            raise FileNotFoundError(msg)
+        return _parse_output_eventlog(
+            eventlog["guest.output"], tasks, labelio, log_stderr_level
+        )
+
+    stream_dict = {"stdout": [], "stderr": [], "log": []}
+    tasks = Taskset(tasks)
+
+    #  Now block until output eventlog is ready:
+    try:
+        event_wait(flux_handle, jobid, "shell.init", "guest.exec.eventlog")
+    except OSError:
+        #  event_wait() could fail due to a job exception before the 'start'
+        #  event. Check for a fatal job exception in the job eventlog
+        #  and raise it as such if so. O/w, reraise the event_wait() exception.
+        #
+        event = event_wait(flux_handle, jobid, "exception")
+        if event is not None and event.context["severity"] == 0:
+            raise JobException(event) from None
+        raise
+
+    #  Output eventlog is ready, synchronously gather all output
+    for event in event_watch(flux_handle, jobid, "guest.output"):
+        _output_eventlog_entry_decode(
+            event, stream_dict, tasks, labelio, log_stderr_level
+        )
+
+    #  Join lines and return JobOutput result
+    results = ["".join(stream_dict.get(k)) for k in ("stdout", "stderr", "log")]
+    return JobOutput(*results)
+
+
+class JobOutputEventWatch(FutureExt):
+    """
+    A class for watching job output events.
+
+    See output_watch_events_async() for full documentation.
+    """
+
+    def __init__(self, flux_handle, jobid, labelio=False, nowait=False):
+        self.labelio = labelio
+        self.nowait = nowait
+
+        #  Capture when 'start' and 'finish' events have been posted to
+        #  the main eventlog:
+        self.started = False
+        self.finished = False
+
+        #  Capture when the end of the output eventlog has been reached
+        #  (indicated by None returned from watching the eventlog)
+        self.closed = False
+        super().__init__(self._watch_init, JobID(jobid), flux_handle=flux_handle)
+
+    def get_event(self, autoreset=True):
+        event = self.get()
+        if event is None:
+            return None
+        if autoreset:
+            self.reset()
+        return _parse_output_eventlog_entry(event, labelio=self.labelio)
+
+    def _watch_output(self, future):
+        #  Watch output events and propagate to output_watch_future.
+        #
+        #  Note: handle and propagate OSError from future.get_event() for
+        #  reasons noted in wait_for_start_event() below.
+        #
+        event = None
+        try:
+            event = future.get_event(autoreset=False)
+            # Note: do not fulfill
```

Comment:
"do not fulfill immediately" is what is meant?

Reply:
Hm, I think the rest of the comment must have gotten accidentally chopped off at some point. I can't really remember what was being said here, but there's a nice long comment in the following conditional that seems to explain the situation so I'll just delete this stray comment.